### PR TITLE
KAFKA-18057: Add share partition redundant offset manager.

### PR DESCRIPTION
**Pull request by smjn**
* The share coordinator writes state to the internal topic `__share_group_state`. This topic is not compacted.
* To prevent the size of partitions of the topic to become unwieldy, we would want to prune this topic periodically.
* Since the disk log can only supports prefix based delete, we require the knowledge of the last redundant offset below which the partition can be truncated. This PR introduces a class which manages the state to calculate the said offset.
* Tests have also been added.

### CODE REVIEW

File: share-coordinator/src/main/java/org/apache/kafka/coordinator/share/SharePartitionOffsetManager.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+package org.apache.kafka.coordinator.share;
+
+import org.apache.kafka.server.share.SharePartitionKey;
+import org.apache.kafka.timeline.SnapshotRegistry;
+import org.apache.kafka.timeline.TimelineHashMap;
+
+import java.util.Objects;
+import java.util.Optional;
+
+public class SharePartitionOffsetManager {
```
Comment (AndrewJSchofield) [inline share-coordinator/src/main/java/org/apache/kafka/coordinator/share/SharePartitionOffsetManager.java:34]: This needs a class-level comment saying what its purpose is.

File: share-coordinator/src/main/java/org/apache/kafka/coordinator/share/SharePartitionOffsetManager.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        offsets = new TimelineHashMap<>(snapshotRegistry, 0);
+    }
+
+    /**
+     * Method updates internal state with the supplied offset for the provided
+     * share partition key. It then calculates the minimum offset, if possible,
+     * below which all offsets are redundant. This value is then returned as
+     * an optional.
+     * <p>
+     * The value returned is exclusive, in that all offsets below it but non including
```
Comment (AndrewJSchofield) [inline share-coordinator/src/main/java/org/apache/kafka/coordinator/share/SharePartitionOffsetManager.java]: nit: "not including"

File: share-coordinator/src/main/java/org/apache/kafka/coordinator/share/SharePartitionOffsetManager.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        minOffset = Math.min(minOffset, offset);
+        offsets.put(key, offset);
+
+        Optional<Long> deleteTillOffset = lastRedundantOffset();
+        deleteTillOffset.ifPresent(off -> minOffset = off);
+        return deleteTillOffset;
+    }
+
+    private Optional<Long> lastRedundantOffset() {
+        long soFar = Long.MAX_VALUE;
```
Comment (AndrewJSchofield) [inline share-coordinator/src/main/java/org/apache/kafka/coordinator/share/SharePartitionOffsetManager.java:72]: I would expect this method to return `Optional.empty()` if there are no offsets in the map.

File: share-coordinator/src/main/java/org/apache/kafka/coordinator/share/SharePartitionOffsetManager.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+            soFar = Math.min(soFar, offset);
+
+            // minOffset represents the smallest necessary offset
+            // and if soFar equals it, we cannot proceed. This can happen
+            // if a share partition key hasn't had records written for a while
+            // For example,
+            // <p>
+            // key1:1
+            // key2:2 4 6
+            // key2:3 5 7
```
Comment (AndrewJSchofield) [inline share-coordinator/src/main/java/org/apache/kafka/coordinator/share/SharePartitionOffsetManager.java]: nit: Should this line be "key3"?

File: share-coordinator/src/main/java/org/apache/kafka/coordinator/share/SharePartitionOffsetManager.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+public class SharePartitionOffsetManager {
+
+    // map to store share partition key => current partition offset
+    // being written
+    private final TimelineHashMap<SharePartitionKey, Long> offsets;
+
+    // minimum offset representing the smallest necessary offset (non-redundant)
+    // across the internal partition
+    private long minOffset = Long.MAX_VALUE;
```
Comment (AndrewJSchofield) [inline share-coordinator/src/main/java/org/apache/kafka/coordinator/share/SharePartitionOffsetManager.java:42]: I'm surprised that 0 is not a better starting value.

**Review by AndrewJSchofield — CHANGES_REQUESTED** [review]

File: share-coordinator/src/main/java/org/apache/kafka/coordinator/share/SharePartitionOffsetManager.java (hunk 5 above)
Comment (smjn) [reply share-coordinator/src/main/java/org/apache/kafka/coordinator/share/SharePartitionOffsetManager.java:42]: This variable is used as arg in call to `Math.min` - 0 might not be suitable for that.

File: share-coordinator/src/main/java/org/apache/kafka/coordinator/share/SharePartitionOffsetManager.java (hunk 3 above)
Comment (smjn) [reply share-coordinator/src/main/java/org/apache/kafka/coordinator/share/SharePartitionOffsetManager.java:72]: I will add but having no offsets in map is not possible in the current state as this is a private method which is called by `updateState` which will for sure have some value in the map before making a call to this method.

File: share-coordinator/src/main/java/org/apache/kafka/coordinator/share/SharePartitionOffsetManager.java (hunk 5 above)
Comment (AndrewJSchofield) [reply share-coordinator/src/main/java/org/apache/kafka/coordinator/share/SharePartitionOffsetManager.java:42]: Of course. Long.MAX_VALUE would be a catastrophic offset to delete from 😨

File: share-coordinator/src/main/java/org/apache/kafka/coordinator/share/SharePartitionOffsetManager.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+ * per share partition key.
+ * It calculates the minimum offset globally up to which the records
+ * in the internal partition are redundant i.e. they have been overridden
+ * by newer records.
+ */
+public class SharePartitionOffsetManager {
+
+    // map to store share partition key => current partition offset
+    // being written
+    private final TimelineHashMap<SharePartitionKey, Long> offsets;
```
Comment (AndrewJSchofield) [inline share-coordinator/src/main/java/org/apache/kafka/coordinator/share/SharePartitionOffsetManager.java:38]: It's not evident to me why this needs to be a TimelineHashMap yet, but this code is still just the first piece of the puzzle.

File: share-coordinator/src/test/java/org/apache/kafka/coordinator/share/SharePartitionOffsetManagerTest.java [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+            this.tuples = tuples;
+            this.shouldRun = shouldRun;
+        }
+    }
+
+    static Stream<ShareOffsetTestHolder> generateNoRedundantStateCases() {
+        return Stream.of(
+                new ShareOffsetTestHolder(
+                        "no redundant state single key",
+                        Collections.singletonList(
```
Comment (AndrewJSchofield) [inline share-coordinator/src/test/java/org/apache/kafka/coordinator/share/SharePartitionOffsetManagerTest.java]: List.of?

File: share-coordinator/src/test/java/org/apache/kafka/coordinator/share/SharePartitionOffsetManagerTest.java [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+                                ShareOffsetTestHolder.TestTuple.instance(KEY1, 10L, Optional.empty())
+                        )
+                ),
+
+                new ShareOffsetTestHolder(
+                        "no redundant state multiple keys",
+                        Arrays.asList(
```
Comment (AndrewJSchofield) [inline share-coordinator/src/test/java/org/apache/kafka/coordinator/share/SharePartitionOffsetManagerTest.java]: List.of?

File: share-coordinator/src/test/java/org/apache/kafka/coordinator/share/SharePartitionOffsetManagerTest.java [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+            }
+
+            static TestTuple instance(SharePartitionKey key, long offset, Optional<Long> expectedOffset) {
+                return new TestTuple(key, offset, expectedOffset);
+            }
+        }
+
+        private final String testName;
+        private final List<TestTuple> tuples;
+        private final boolean shouldRun;
```
Comment (AndrewJSchofield) [inline share-coordinator/src/test/java/org/apache/kafka/coordinator/share/SharePartitionOffsetManagerTest.java:86]: This seems unnecessary at the moment.

**Review by AndrewJSchofield — APPROVED** [review]
I might have further comments once I see how it is used. Looks good to me as a starting point.

File: share-coordinator/src/test/java/org/apache/kafka/coordinator/share/SharePartitionOffsetManagerTest.java (hunk 9 above)
Comment (smjn) [reply share-coordinator/src/test/java/org/apache/kafka/coordinator/share/SharePartitionOffsetManagerTest.java:86]: It makes very easy to run a single test among the stream. Helpful in debugging

**Pull request closed without merging**